File: android/src/main/java/com/trustwalletcorereactnative/TrustWalletCoreCoin.kt
package com.trustwalletcorereactnative

import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.trustwalletcorereactnative.wallet_core.coin.Coin
import com.trustwalletcorereactnative.wallet_core.error.CEError
import com.trustwalletcorereactnative.wallet_core.error.CError

class TrustWalletCoreCoin(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private fun resolveError(e : Exception,promise: Promise){
        if(e is CError){
            promise.reject(e.getCode(),e.message , null)
        }else{
            promise.reject(CEError.Unknown.code,e.message , null)
        }
    }

    private fun convertReadableMapToMap(readableMap: ReadableMap): Map<String, Any> {
        val map: MutableMap<String, Any> = HashMap()
        val iterator = readableMap.keySetIterator()

        while (iterator.hasNextKey()) {
            val key = iterator.nextKey()

            when (readableMap.getType(key)) {
                ReadableType.Boolean -> map[key] = readableMap.getBoolean(key)
                ReadableType.Number -> map[key] = readableMap.getDouble(key)
                ReadableType.String -> map[key] = readableMap.getString(key)!!
                ReadableType.Map -> map[key] = convertReadableMapToMap(readableMap.getMap(key)!!)
                ReadableType.Array -> map[key] =
                    convertReadableArrayToList(readableMap.getArray(key)!!)

                else -> throw IllegalArgumentException("Unsupported type")
            }
        }
        return map
    }

    private fun convertReadableArrayToList(readableArray: ReadableArray): List<Any> {
        val list: MutableList<Any> = ArrayList()
        for (i in 0 until readableArray.size()) {
            when (readableArray.getType(i)) {
                ReadableType.Boolean -> list.add(readableArray.getBoolean(i))
                ReadableType.Number -> list.add(readableArray.getDouble(i))
                ReadableType.String -> list.add(readableArray.getString(i))
                ReadableType.Map -> list.add(convertReadableMapToMap(readableArray.getMap(i)))
                ReadableType.Array -> list.add(convertReadableArrayToList(readableArray.getArray(i)))
                else -> throw IllegalArgumentException("Unsupported type")
            }
        }
        return list
    }



    override fun getName(): String {
        return NAME
    }

    @ReactMethod
    fun getAddress(mnemonic: String, passPhrase: String, coinType: Int, promise: Promise) {
        try{
            val coin = Coin.createCoinByteType(coinType)

            val address = coin.getAddress(mnemonic, passPhrase)

            promise.resolve(address)
        }catch (e : Exception){
            resolveError(e,promise)
        }
    }

    @ReactMethod
    fun getSeed(mnemonic: String, passPhrase: String, coinType: Int, promise: Promise) {
        try{
            val coin = Coin.createCoinByteType(coinType)

            val seed = coin.getSeed(mnemonic, passPhrase)

            promise.resolve(seed)
        }catch (e : Exception){
            resolveError(e,promise)
        }
    }

    @ReactMethod
    fun getHexPrivateKey(mnemonic: String, passPhrase: String, coinType: Int, promise: Promise) {
        try{
            val coin = Coin.createCoinByteType(coinType)

            val privateKey = coin.getHexPrivateKey(mnemonic, passPhrase)

            promise.resolve(privateKey)
        }catch (e : Exception){
            resolveError(e,promise)
        }
    }

    @ReactMethod
    fun getRawPrivateKey(mnemonic: String, passPhrase: String, coinType: Int, promise: Promise) {
        try{
            val coin = Coin.createCoinByteType(coinType)

            val privateKey = coin.getRawPrivateKey(mnemonic, passPhrase)

            val array = Arguments.createArray()

            for (b in privateKey){
                array.pushInt(b.toInt())
            }

            promise.resolve(array)
        }catch (e : Exception){
            resolveError(e, promise)
        }
    }

    @ReactMethod
    fun getHexPublicKey(mnemonic: String, passPhrase: String, coinType: Int, promise: Promise) {
        try{
            val coin = Coin.createCoinByteType(coinType)

            val publicKey = coin.getHexPublicKey(mnemonic, passPhrase)

            promise.resolve(publicKey)
        }catch (e : Exception){
            resolveError(e,promise)
        }
    }

    @ReactMethod
    fun getRawPublicKey(mnemonic: String, passPhrase: String, coinType: Int, promise: Promise) {
        try {
            val coin = Coin.createCoinByteType(coinType)

            val array = Arguments.createArray()

            val publicKey = coin.getRawPublicKey(mnemonic, passPhrase)

            for (p in publicKey){
                array.pushInt(p.toInt())
            }

            promise.resolve(array)
        }catch (e: Exception){
            resolveError(e,promise)
        }
    }

    @ReactMethod
    fun sign(mnemonic: String, passPhrase: String, coinType: Int, message: String, promise: Promise){
        try {
            val coin = Coin.createCoinByteType(coinType)

            val sig = coin.sign(mnemonic,passPhrase,message)

            promise.resolve(sig)
        }catch (e: Exception){
            resolveError(e,promise)
        }
    }

    @ReactMethod
    fun signTransaction(mnemonic: String, passPhrase: String, coinType: Int, tx: ReadableMap, promise: Promise){
        try{
            val coin = Coin.createCoinByteType(coinType)

            val signed = coin.signTransaction(convertReadableMapToMap(tx),mnemonic,passPhrase)

            promise.resolve(signed)
        }catch (e : Exception){
            resolveError(e,promise)
        }
    }

    @ReactMethod
    fun signTransactionWithPrivateKey(privateKey : String,coinType: Int, tx: ReadableMap, promise: Promise){
        try {
            val coin = Coin.createCoinByteType(coinType)

            val signed = coin.signTransaction(convertReadableMapToMap(tx),privateKey)

            promise.resolve(signed)
        }catch (e : Exception){
            resolveError(e,promise)
        }
    }

//    @ReactMethod
//    fun multiSignTransaction(mnemonic: String, passphrase: String, tx: ReadableMap, privateKeys: ArrayList<ByteArray>,coinType: Int, promise: Promise){
//        val coin = Coin.createCoinByteType(coinType)
//
//        val array = Arguments.createArray()
//
//        val signed = coin.multiSignTransaction(mnemonic,passphrase,convertReadableMapToMap(tx),privateKeys)
//
//        for (s in signed){
//            val arr = Arguments.createArray()
//
//            for (raw in s){
//                arr.pushInt(raw.toInt())
//            }
//
//            array.pushArray(arr)
//        }
//
//        promise.resolve(array)
//    }

    companion object {
        const val NAME = "TrustWalletCoreCoin"
    }
}

File: ios/TrustWalletCoreCoin.swift
@objc(TrustWalletCoreCoin)
class TrustWalletCoreCoin: NSObject {
    
    func convertDataToReactNativeResult(data : Data) -> [Int32]{
        let intCount = data.count / MemoryLayout<Int32>.size
        let intArrayFromData = data.withUnsafeBytes { ptr in
            Array(UnsafeBufferPointer<Int32>(start: ptr.bindMemory(to: Int32.self).baseAddress!, count: intCount))
        }

        return intArrayFromData
    }

    @objc
    func getAddress(_ mnemonic: String, passPharse : String, coinType : Int , resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        do {
            let coin = try Coin.createCoinByType(coinType: coinType)
            
            let address = try coin.getAddress(mnemonic: mnemonic, passPharse: passPharse)
            
            resolve(address)
        }catch let error as CError{
            reject(error.toReactCode(),error.toReactException(),nil)
        }catch {
            reject(CError.Unknown.toReactCode(),error.localizedDescription,nil)
        }
    }
    
    @objc
    func getSeed(_ mnemonic: String, passPharse : String, coinType : Int , resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        do {
            let coin = try Coin.createCoinByType(coinType: coinType)
            
            let seed = try coin.getSeed(mnemonic: mnemonic, passPharse: passPharse)
            
            let array = convertDataToReactNativeResult(data: seed)
            
            resolve(array)
        }catch let error as CError{
            reject(error.toReactCode(),error.toReactException(),nil)
        }catch {
            reject(CError.Unknown.toReactCode(),error.localizedDescription,nil)
        }
    }
    
    @objc
    func getHexPrivateKey(_ mnemonic: String, passPharse : String, coinType : Int , resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        do {
            let coin = try Coin.createCoinByType(coinType: coinType)
            
            let privateKey = try coin.getHexPrivateKey(mnemonic: mnemonic, passPharse: passPharse)
            
            resolve(privateKey)
        }catch let error as CError{
            reject(error.toReactCode(),error.toReactException(),nil)
        }catch {
            reject(CError.Unknown.toReactCode(),error.localizedDescription,nil)
        }
    }
    
    @objc
    func getRawPrivateKey(_ mnemonic: String, passPharse : String, coinType : Int , resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        do {
            let coin = try Coin.createCoinByType(coinType: coinType)
            
            let privateKey = try coin.getRawPrivateKey(mnemonic: mnemonic, passPharse: passPharse)
            
            let array = convertDataToReactNativeResult(data: privateKey)
            
            resolve(array)
        }catch let error as CError{
            reject(error.toReactCode(),error.toReactException(),nil)
        }catch {
            reject(CError.Unknown.toReactCode(),error.localizedDescription,nil)
        }
    }
    
    @objc
    func getHexPublicKey(_ mnemonic: String, passPharse : String, coinType : Int , resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        do {
            let coin = try Coin.createCoinByType(coinType: coinType)
            
            let publicKey = try coin.getHexPulicKey(mnemonic: mnemonic, passPharse: passPharse)
            
            resolve(publicKey)
        }catch let error as CError{
            reject(error.toReactCode(),error.toReactException(),nil)
        }catch {
            reject(CError.Unknown.toReactCode(),error.localizedDescription,nil)
        }
    }
    
    @objc
    func getRawPublicKey(_ mnemonic: String, passPharse : String, coinType : Int , resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        do {
            let coin = try Coin.createCoinByType(coinType: coinType)
            
            let publicKey = try coin.getRawPublicKey(mnemonic: mnemonic, passPharse: passPharse)
            
            let array = convertDataToReactNativeResult(data: publicKey)
            
            resolve(array)
        }catch let error as CError{
            reject(error.toReactCode(),error.toReactException(),nil)
        }catch {
            reject(CError.Unknown.toReactCode(),error.localizedDescription,nil)
        }
    }
    
    
    @objc
    func sign(_ mnemonic: String, passPharse : String, coinType : Int ,message : String, resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        do {
            let coin = try Coin.createCoinByType(coinType: coinType)
            
            let sig = try coin.sign(mnemonic: mnemonic, passPharse: passPharse, message: message)
            
            resolve(sig)
        }catch let error as CError{
            reject(error.toReactCode(),error.toReactException(),nil)
        }catch {
            reject(CError.Unknown.toReactCode(),error.localizedDescription,nil)
        }
    }
    
    @objc
    func signTransaction(_ mnemonic: String, passPharse : String, coinType : Int , tx: Dictionary<String, Any> ,resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock){
        do {
            let coin = try Coin.createCoinByType(coinType: coinType)
            
            let tx = try coin.signTransaction(mnemonic: mnemonic, passPharse: passPharse, tx: tx)
            
            resolve(tx)
        }catch let error as CError{
            reject(error.toReactCode(),error.toReactException(),nil)
        }catch {
            reject(CError.Unknown.toReactCode(),error.localizedDescription,nil)
        }
    }
    
    @objc
    func signTransactionWithPrivateKey(_ privateKey : String,coinType : Int , tx: Dictionary<String, Any> ,resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock){
        do {
            let coin = try Coin.createCoinByType(coinType: coinType)
            
            let tx = try coin.signTransaction(privateKey: privateKey, tx: tx)
            
            resolve(tx)
        }catch let error as CError{
            reject(error.toReactCode(),error.toReactException(),nil)
        }catch {
            reject(CError.Unknown.toReactCode(),error.localizedDescription,nil)
        }
    }
}
